haskell/nix.bzl

```diff
@@ -48,3 +48,119 @@ targets()
+        )
+    repository_ctx.file("BUILD", build_file_content)
+
+bundle = repository_rule(
+    attrs = {
+        "packages": attr.string_list(),
+        "base_repository": attr.string(),
+    },
+    implementation = _bundle_impl,
+)
+
+def haskell_nixpkgs_packages(name, base_attribute_path, packages, *args, **kwargs):
+    """Import a set of haskell packages from nixpkgs.
+
+    This takes as input the same arguments as `nixpkgs_package`, expecting the
+    `attribute_path` to resolve to a set of haskell packages (such as
+    `haskellPackages` or `haskell.packages.ghc822`) preprocessed by the
+    `genBazelBuild` function. It also takes as input a list of packages to
+    import (which can be generated by the `gen_packages_list` function).
+
+    Args:
+
+      base_attribute_path: The attribute path in the nix expression pointing to
+          the `haskellPackages` set
+      packages: The list of packages to generate a definition for
+
+    Example:
+
+      In `haskellPackages.nix`:
+
+      ```nix
+      with import <nixpkgs> {};
+
+      let wrapPackages = callPackage <bazel_haskell_wrapper> { }; in
+      { haskellPackages = genBazelBuild haskell.packages.ghc822; }
+      ```
+
+      In your `WORKSPACE`
+
+      ```bazel
+      # Define a nix repository to fetch the packages from
+      load("@io_tweag_rules_nixpkgs//nixpkgs:nixpkgs.bzl",
+          "nixpkgs_git_repository")
+      nixpkgs_git_repository(
+          name = "nixpkgs",
+          revision = "9a787af6bc75a19ac9f02077ade58ddc248e674a",
+      )
+
+      load("@io_tweag_rules_haskell//haskell:nix.bzl",
+          "haskell_nixpkgs_packages",
+          "gen_packages_list")
+
+      # Generate a list of all the available haskell packages
+      gen_packages_list(
+          name = "hackage-packages",
+          repositories = {"@nixpkgs": "nixpkgs"},
+          nix_file = "//haskellPackages.nix",
+          base_attribute_path = "haskellPackages",
+      )
+      load("@hackage-packages//:all-haskell-packages.bzl", "packages")
+
+      # Define the set of haskell packages
+      haskell_nixpkgs_packages(
+          name = "hackage",
+          packages = packages,
+          repositories = { "@nixpkgs": "nixpkgs" },
+          nix_file = "haskellPackages.nix",
+          base_attribute_path = "haskellPackages",
+      )
+      ```
+
+      Then in your `BUILD` files, you can access to the whole of hackage as
+      `@hackage//:{your-package-name}`
+    """
+    for package in packages:
+        haskell_nixpkgs_package(
+            name = name + "-" + package,
+            attribute_path = base_attribute_path + "." + package,
+            *args,
+            **kwargs
+        )
+    bundle(
+        name = name,
+        packages = packages,
+        base_repository = name,
+    )
+
+def gen_packages_list(name, base_attribute_path, repositories = {}, *args, **kwargs):
+    """Generate a list of all the available haskell packages"""
```

Comment:
I think this needs a bit of documentation itself, for what the attributes represent.

Reply:
I also think its name should be reconsidered so that it is namespaced to `rules_haskell` appropriately. E.g. `haskell_...`. That said, is there a use case where I wouldn't immediately tee this up into `haskell_nixpkgs_packages`? If not, we could combine them into a single function which would arguably be a lower learning curve?

Reply:
Agreed for the name and the documentation. Would `haskell_nixpkgs_package_list` be OK? That doesn't sound awesome but I can't come up with anything better.

For the merge with `haskell_nixpkgs_packages`, I'd like to have it, but `load` statements can't be placed in the body of functions in bazel, so to the top of my knowledge it isn't possible.